modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala

```diff
@@ -219,4 +242,51 @@ object Enrich {
         .getBytes(UTF_8)
     } else originalBytes
   }
+
+  /**
+   * This is the machinery needed to make sure that no chunk is sunk without
+   * being checkpointed when the app terminates.
+   *
+   * The stream runs on a separate fiber so that we can manually handle SIGINT.
+   *
+   * We use a queue as a level of indirection between the stream of enriched events and the sink + checkpointing.
+   * When we receive a SIGINT or exception then we terminate the fiber by pushing a `None` to the queue.
+   *
+   * The stream is only cancelled after the sink + checkpointing have been allowed to finish cleanly.
+   * We must not terminate the source any earlier, because this would shutdown the kinesis scheduler too early,
+   * and then we would not be able to checkpoint the outstanding records.
+   */
+  private def runWithShutdown[F[_]: Concurrent: Sync: Timer, A](
+    enriched: Stream[F, List[(A, Result)]],
+    sinkAndCheckpoint: Pipe[F, List[(A, Result)], Unit]
+  ): F[Unit] =
+    Queue.synchronousNoneTerminated[F, List[(A, Result)]].flatMap { queue =>
+      queue.dequeue
+        .through(sinkAndCheckpoint)
+        .concurrently(enriched.evalMap(x => queue.enqueue1(Some(x))).onFinalize(queue.enqueue1(None)))
+        .compile
+        .drain
+        .start
+        .bracketCase(_.join) {
+          case (_, ExitCase.Completed) =>
+            // The source has completed "naturally", e.g. processed all input files in the directory
+            Sync[F].unit
+          case (fiber, ExitCase.Canceled) =>
+            // SIGINT received. We wait for the enriched events already in the queue to get sunk and checkpointed
+            terminateStream(queue, fiber)
+          case (fiber, ExitCase.Error(e)) =>
+            // Runtime exception in the stream of enriched events.
+            // We wait for the enriched events already in the queue to get sunk and checkpointed.
+            // We then raise the original exception
+            terminateStream(queue, fiber).attempt *> Logger[F].error(e)("Unexpected error in enrich") *> Sync[F].raiseError(e)
```

Comment:
Hey @benjben , I was lucky enough to have a pipeline that threw an exception!  So I was able to see what logs we get on error:

```
[pool-1-thread-4] WARN com.snowplowanalytics.snowplow.enrich.common.fs2.Enrich - Terminating enrich stream. Waiting 5 minutes for it to complete                                                                                            
[pool-1-thread-8] ERROR com.snowplowanalytics.snowplow.enrich.common.fs2.Run - The Enrich job has stopped                                                                                                                                   
[pool-1-thread-8] INFO org.http4s.client.PoolManager - Shutting down connection pool: curAllocated=1 idleQueues.size=1 waitQueue.size=0 maxWaitQueueLimit=256 closed=false                                                                  
java.lang.NullPointerException                                                                                                                                                                                                              
        at io.circe.Printer$PrintingFolder.onString(Printer.scala:309)                                                                                                                                                                      
        at io.circe.Printer$PrintingFolder.onString(Printer.scala:292)   
  -- etc
  ```

Do you think it's a possible problem that it does not print the error message until after it has shut down the stream?  I was thinking if the orchestration (e.g. docker/kubernetes) sends a forceful kill signal while Enrich is still trying to terminate, then you never see the log message at all.

In BigQuery loader I made it so that it [logs the error message before trying to terminate the stream](https://github.com/snowplow-incubator/snowplow-bigquery-loader/pull/250/files#diff-a92f3fcd4f8171febee7c6990a044ba161df3cb662414046bebe3772e7dabbaeR111).  So if something goes wrong after that, then at least we know the error message has been printed.

It's a small detail, sorry for nitpicking.

Reply:
Yeah good point, I'll put the log before.